[Index Management] Action button inside ILM banner is rendered without text

damian-polewski opened the pull request:
Closes #223369

## Summary

This PR fixes an issue where action button was rendered without text inside the ILM Banner on the Index Management page.

### Context

Inside the ILM Banner we show the button that allows users to retry lifecycle step in case of any lifecycle errors.

<img width="1918" height="932" alt="Screenshot 2025-07-17 at 10 49 48" src="https://github.com/user-attachments/assets/731f40d3-5722-43de-a442-37660dac3adb" />

The problem arises when we have some other lifecycle error e.g. we assign lifecycle policy that doesn't exist to an index. In this case `retryLifecycleActionExtension()` returns `null` because the index is missing `ilm.failed_step` and so the `action.buttonLabel` is `undefined` (as well as other keys in `action` object).

<img width="1918" height="933" alt="Screenshot 2025-07-17 at 10 51 38" src="https://github.com/user-attachments/assets/4cc0c877-d14b-42ae-bad7-8e31176acc9d" />
<img width="1917" height="934" alt="Screenshot 2025-07-17 at 10 51 57" src="https://github.com/user-attachments/assets/2eec89b0-161a-49c0-aa67-1f9ade4c19e5" />

To fix this we need make sure we only add the `action` object to banner when retry action is available for all indices with lifecycle errors (when `retryLifecycleActionExtension()` return an object instead of `null`).
This assures that the button is not rendered if we cannot apply same action to all indices with lifecycle errors.

### How to test

Create an index and assign to it a non-existing lifecycle policy.
```
PUT my_index
{
    "settings": {
        "index.lifecycle.name": "test"
    }
}
```

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

<!--ONMERGE {"backportTargets":["8.17","8.18","8.19","9.0","9.1"]} ONMERGE-->

damian-polewski committed:
fix: ilm banner action button issue

SoniaSanzV reviewed (commented):
Great work Damian! In addition of what you did, I would suggest filtering out the indices without failed step in `retryLifecycleActionExtension` so the button can be available when we have more than one failed indices and one of them can be retried. Something like:

```
export const retryLifecycleActionExtension = ({ indices }: { indices: Index[] }) => {
  const indicesWithFailedStep = indices.filter((index) => {
    return index.ilm?.managed && index.ilm.failed_step;
  });
  if (!indicesWithFailedStep.length) {
    return null;
  }
  const indexNames = indicesWithFailedStep.map(({ name }: Index) => name);
  return {
    requestMethod: retryLifecycleForIndex,
    icon: 'play',
    indexNames: [indexNames],
    buttonLabel: i18n.translate('xpack.indexLifecycleMgmt.retryIndexLifecycleActionButtonLabel', {
      defaultMessage: 'Retry lifecycle step',
    }),
    successMessage: i18n.translate(
      'xpack.indexLifecycleMgmt.retryIndexLifecycleAction.retriedLifecycleMessage',
      {
        defaultMessage: 'Called retry lifecycle step for: {indexNames}',
        values: { indexNames: indexNames.map((indexName: string) => `"${indexName}"`).join(', ') },
      }
    ),
  };
};
```

damian-polewski committed:
Merge branch 'main' into index_management/ilm_banner_action_undefined

damian-polewski committed:
Merge branch 'main' into index_management/ilm_banner_action_undefined

damian-polewski committed:
fix: cr changes

damian-polewski commented:
Hi @SoniaSanzV, thank you for your feedback!
I've changed `retryLifecycleActionExtension` as you suggested and adjusted unit tests.

I also changed button copy `Retry lifecycle step` to `Retry failed lifecycle step` to be more clear that we are only running retry action for indices with failed step lifecycle errors. @elastic/docs could you let me know what do you think about this copy change?

| Lifecycle error | Before | After |
| -------- | ------- | ------- |
| failed step error  | <img width="1916" height="932" alt="Screenshot 2025-07-21 at 09 05 13" src="https://github.com/user-attachments/assets/91475e8e-c70d-4774-a3e5-38a32fd7443a" /> |  |
| non existent policy error  | <img width="1915" height="932" alt="Screenshot 2025-07-21 at 09 06 03" src="https://github.com/user-attachments/assets/24847b94-d57c-4a36-b72e-09e0e94a6fe4" /> | <img width="1916" height="934" alt="Screenshot 2025-07-21 at 07 49 14" src="https://github.com/user-attachments/assets/671f019a-8688-440d-9647-6694dd50a7c2" /> |
| non existent policy error and failed step error | <img width="1918" height="935" alt="Screenshot 2025-07-21 at 09 03 48" src="https://github.com/user-attachments/assets/a775056a-665a-4c78-8d1e-af6d2355ecb6" /> | <img width="1916" height="933" alt="Screenshot 2025-07-21 at 07 41 53" src="https://github.com/user-attachments/assets/0f7caebf-aa2c-497e-896c-d91ab99a298d" /> |

florent-leborgne reviewed (approved):
Copy change LGTM

SoniaSanzV reviewed (approved):
thank you for addressing my feedback. Tested locally, lgtm :)

kibanamachine commented:
Starting backport for target branches: 8.17, 8.18, 8.19, 9.0, 9.1

https://github.com/elastic/kibana/actions/runs/16413605940

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228734">](https://github.com/elastic/kibana/pull/228734)|
|✅|8.18|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228735">](https://github.com/elastic/kibana/pull/228735)|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228736">](https://github.com/elastic/kibana/pull/228736)|
|✅|9.0|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228737">](https://github.com/elastic/kibana/pull/228737)|
|✅|9.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/228738">](https://github.com/elastic/kibana/pull/228738)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->
